daffodil-io/src/main/scala/org/apache/daffodil/io/InputSourceDataInputStream.scala

```diff
@@ -117,6 +118,57 @@ final class InputSourceDataInputStream private (val inputSource: InputSource)
   @inline
   override final def bitLimit0b: MaybeULong = cst.bitLimit0b
 
+  /**
+   * Tells us if the underlying input source has detected end-of-data
+   * (the read(...) call returned -1.
+   *
+   * But this does NOT tell us we are positioned at the end, only whether
+   * in the course of reading, we encountered the end of data. If we
+   * backtracked we could have seen the end of data, but backed up in
+   * the data to an earlier position.
+   */
+  def hasReachedEndOfData: Boolean = inputSource.hasReachedEndOfData
+
+  /**
+   * Determine if we're positioned at the end of data without
+   * doing any additional blocking operation such as reading more
+   * data to test if there is any.
+   *
+   * This depends on the underlying inputSource keeping track of
+   * whether it has previously hit the end of data or not.
+   *
+   * @return
+   */
+  final def isAtEnd(): Boolean = {
+    if (bitLimit0b.isDefined) {
+      bitPos0b == bitLimit0b.get
+    } else {
+      // There is no bit limit, so whether we're at the end
+      // requires that we've hit then end-of-data on a read
+      // operation, AND that we've not backtracked since then,
+      // meaning we're positioned at the end of the last byte
+      // read.
+      hasReachedEndOfData && {
+        //
+        // No real input source can end in the middle of a byte.
+        // The only way that can happen is with the bitLimit having been
+        // set, and we've already dealt with that.
+        //
+        // So we just have to know that we're positioned at the
+        // end of the data.
+        //
+        val inputStreamBytePos0b = inputSource.position()
+        val isAtFinalByte = bytePos0b == inputStreamBytePos0b
```

Comment:
I thought bytePos0b and inputSource.position() were basically always kept in sync? When we change bytePos0b, don't we also change the InputSource byte position, even when we backtrack. Seems like this is always going to be true, even if we've backtracked. Does the InputSource need a isAtEndOfData since only it really knows whre the last byte in the InputSource is?

Reply:
Ok. I get your reasoning. 
Not sure the fix for this. (This code is now gone, but the question is still relevant.)
The scenario is we parsed forward, possibly ran into EOD, possibly not, but then we backtracked, and are no longer at the extreme end of where we reached with the parser. 
We need to recognize that case, and DataInputStream needs a way to support this, perhaps by keeping track of the bytePos of the last byte of the last filled bucket?